File: src/search/chunker.ts
/**
 * Chunker
 * Split documents and tasks into semantic chunks for embedding
 */

import type { Task } from "@models/task";
import type {
	Chunk,
	ChunkResult,
	DocChunk,
	DocMetadata,
	EmbeddingModel,
	MarkdownHeading,
	TaskChunk,
	TaskChunkableField,
} from "./types";
import { EMBEDDING_MODELS, getTaskFieldContent } from "./types";

/**
 * Simple token estimation (approx 4 chars per token for English)
 * This is a rough estimate - actual tokenization depends on the model
 */
export function estimateTokens(text: string): number {
	if (!text) return 0;
	// Rough estimate: ~4 characters per token for English text
	// This is conservative to avoid exceeding limits
	return Math.ceil(text.length / 4);
}

/**
 * Extract headings from markdown content
 */
export function extractHeadings(markdown: string): MarkdownHeading[] {
	const headings: MarkdownHeading[] = [];
	const lines = markdown.split("\n");

	let currentHeading: MarkdownHeading | null = null;
	let contentStartIndex = 0;

	for (let i = 0; i < lines.length; i++) {
		const line = lines[i];
		const headingMatch = line.match(/^(#{1,6})\s+(.+)$/);

		if (headingMatch) {
			// Save previous heading with its content
			if (currentHeading) {
				const contentEndIndex = lines.slice(0, i).join("\n").length;
				currentHeading.content = markdown.substring(currentHeading.startIndex, contentEndIndex).trim();
				currentHeading.endIndex = contentEndIndex;
				headings.push(currentHeading);
			}

			// Start new heading
			const level = headingMatch[1].length;
			const title = headingMatch[2].trim();
			contentStartIndex = lines.slice(0, i).join("\n").length + line.length + 1;

			currentHeading = {
				level,
				title,
				content: "",
				startIndex: contentStartIndex,
				endIndex: markdown.length,
			};
		}
	}

	// Don't forget the last heading
	if (currentHeading) {
		currentHeading.content = markdown.substring(currentHeading.startIndex).trim();
		currentHeading.endIndex = markdown.length;
		headings.push(currentHeading);
	}

	return headings;
}

/**
 * Chunk a document by headings
 */
export function chunkDocument(
	content: string,
	metadata: DocMetadata,
	model: EmbeddingModel = "gte-small",
): ChunkResult {
	const modelConfig = EMBEDDING_MODELS[model];
	const maxTokens = modelConfig.maxTokens;
	const chunks: DocChunk[] = [];
	let totalTokens = 0;

	// First chunk: metadata (title + description)
	const metadataContent = metadata.description ? `${metadata.title}\n\n${metadata.description}` : metadata.title;

	const metadataTokens = estimateTokens(metadataContent);
	chunks.push({
		id: `doc:${metadata.path}:chunk:0`,
		type: "doc",
		docPath: metadata.path,
		section: "# Metadata",
		content: metadataContent,
		tokenCount: metadataTokens,
		metadata: {
			headingLevel: 1,
			position: 0,
		},
	});
	totalTokens += metadataTokens;

	// Extract headings and create chunks
	const headings = extractHeadings(content);

	// Group ## headings as main chunks, ### as sub-chunks
	let position = 1;
	let parentSection: string | undefined;

	for (const heading of headings) {
		// Skip h1 (usually document title, already in metadata)
		if (heading.level === 1) continue;

		const sectionTitle = `${"#".repeat(heading.level)} ${heading.title}`;
		const sectionContent = `${sectionTitle}\n\n${heading.content}`;
		const tokenCount = estimateTokens(sectionContent);

		// If content exceeds max tokens, we need to split it
		if (tokenCount > maxTokens) {
			// Split by paragraphs
			const paragraphs = heading.content.split(/\n\n+/);
			let currentChunkContent = sectionTitle;
			let currentTokenCount = estimateTokens(sectionTitle);

			for (const para of paragraphs) {
				const paraTokens = estimateTokens(para);

				if (currentTokenCount + paraTokens > maxTokens && currentChunkContent !== sectionTitle) {
					// Save current chunk and start new one
					chunks.push({
						id: `doc:${metadata.path}:chunk:${position}`,
						type: "doc",
						docPath: metadata.path,
						section: sectionTitle,
						content: currentChunkContent,
						tokenCount: currentTokenCount,
						metadata: {
							headingLevel: heading.level,
							parentSection: heading.level > 2 ? parentSection : undefined,
							position,
						},
					});
					totalTokens += currentTokenCount;
					position++;

					// Start new chunk (continuation)
					currentChunkContent = `${sectionTitle} (continued)\n\n${para}`;
					currentTokenCount = estimateTokens(currentChunkContent);
				} else {
					currentChunkContent += `\n\n${para}`;
					currentTokenCount += paraTokens;
				}
			}

			// Don't forget remaining content
			if (currentChunkContent !== sectionTitle) {
				chunks.push({
					id: `doc:${metadata.path}:chunk:${position}`,
					type: "doc",
					docPath: metadata.path,
					section: sectionTitle,
					content: currentChunkContent,
					tokenCount: currentTokenCount,
					metadata: {
						headingLevel: heading.level,
						parentSection: heading.level > 2 ? parentSection : undefined,
						position,
					},
				});
				totalTokens += currentTokenCount;
				position++;
			}
		} else {
			// Content fits in one chunk
			chunks.push({
				id: `doc:${metadata.path}:chunk:${position}`,
				type: "doc",
				docPath: metadata.path,
				section: sectionTitle,
				content: sectionContent,
				tokenCount,
				metadata: {
					headingLevel: heading.level,
					parentSection: heading.level > 2 ? parentSection : undefined,
					position,
				},
			});
			totalTokens += tokenCount;
			position++;
		}

		// Track parent section for nested headings
		if (heading.level === 2) {
			parentSection = sectionTitle;
		}
	}

	// If no headings found, treat entire content as one chunk
	if (headings.length === 0 && content.trim()) {
		const contentTokens = estimateTokens(content);
		chunks.push({
			id: `doc:${metadata.path}:chunk:1`,
			type: "doc",
			docPath: metadata.path,
			section: "# Content",
			content: content,
			tokenCount: contentTokens,
			metadata: {
				headingLevel: 1,
				position: 1,
			},
		});
		totalTokens += contentTokens;
	}

	return { chunks, totalTokens };
}

/**
 * Chunk a task by fields
 */
export function chunkTask(task: Task, model: EmbeddingModel = "gte-small"): ChunkResult {
	const chunks: TaskChunk[] = [];
	let totalTokens = 0;

	const fields: TaskChunkableField[] = ["description", "ac", "plan", "notes"];

	for (const field of fields) {
		const content = getTaskFieldContent(task, field);
		if (!content) continue;

		const tokenCount = estimateTokens(content);

		chunks.push({
			id: `task:${task.id}:chunk:${field}`,
			type: "task",
			taskId: task.id,
			field,
			content,
			tokenCount,
			metadata: {
				status: task.status,
				priority: task.priority,
				labels: task.labels,
			},
		});

		totalTokens += tokenCount;
	}

	return { chunks, totalTokens };
}

/**
 * Get all chunkable content from a task as a single string
 * Useful for simple embedding without chunking
 */
export function getTaskSearchableContent(task: Task): string {
	const parts: string[] = [task.title];

	if (task.description) parts.push(task.description);

	if (task.acceptanceCriteria?.length) {
		parts.push(task.acceptanceCriteria.map((ac) => ac.text).join(" "));
	}

	if (task.implementationPlan) parts.push(task.implementationPlan);
	if (task.implementationNotes) parts.push(task.implementationNotes);

	return parts.join("\n\n");
}

File: internal/search/chunker.go
package search

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/howznguyen/knowns/internal/models"
)

// EstimateTokens returns a rough token count (~4 chars per token for English).
func EstimateTokens(text string) int {
	if text == "" {
		return 0
	}
	return (len(text) + 3) / 4 // ceiling division
}

// countTokens uses the tokenizer if available, otherwise falls back to EstimateTokens.
func countTokens(text string, tok Tokenizer) int {
	if tok == nil {
		return EstimateTokens(text)
	}
	out := tok.Encode(text, 999999)
	return len(out.InputIDs)
}

// markdownHeading represents a heading extracted from markdown content.
type markdownHeading struct {
	Level      int
	Title      string
	HeaderPath string // full hierarchy path e.g. "API/Endpoints/GET /users"
	Content    string
	StartIndex int
	EndIndex   int
}

var headingRE = regexp.MustCompile(`^(#{1,6})\s+(.+)$`)

// extractHeadings splits markdown into heading sections.
func extractHeadings(markdown string) []markdownHeading {
	lines := strings.Split(markdown, "\n")
	var headings []markdownHeading
	var current *markdownHeading
	inCodeBlock := false

	// Header stack for tracking full hierarchy path.
	type headerEntry struct {
		level int
		title string
	}
	var headerStack []headerEntry

	buildHeaderPath := func(stack []headerEntry) string {
		if len(stack) == 0 {
			return ""
		}
		parts := make([]string, len(stack))
		for i, e := range stack {
			parts[i] = e.title
		}
		return strings.Join(parts, "/")
	}

	for i, line := range lines {
		// Track fenced code blocks (triple-backtick toggle).
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "```") {
			inCodeBlock = !inCodeBlock
			continue
		}

		// Skip heading detection inside code blocks.
		if inCodeBlock {
			continue
		}

		m := headingRE.FindStringSubmatch(line)
		if m == nil {
			continue
		}

		level := len(m[1])
		title := strings.TrimSpace(m[2])

		// Update header stack: pop entries of equal or higher level.
		for len(headerStack) > 0 && headerStack[len(headerStack)-1].level >= level {
			headerStack = headerStack[:len(headerStack)-1]
		}
		headerStack = append(headerStack, headerEntry{level: level, title: title})
		headerPath := buildHeaderPath(headerStack)

		// Close previous heading.
		if current != nil {
			endIdx := 0
			for j := 0; j < i; j++ {
				endIdx += len(lines[j]) + 1
			}
			current.Content = strings.TrimSpace(markdown[current.StartIndex:endIdx])
			current.EndIndex = endIdx
			headings = append(headings, *current)
		}

		// Compute content start (after this heading line).
		startIdx := 0
		for j := 0; j <= i; j++ {
			startIdx += len(lines[j]) + 1
		}
		if startIdx > len(markdown) {
			startIdx = len(markdown)
		}

		current = &markdownHeading{
			Level:      level,
			Title:      title,
			HeaderPath: headerPath,
			StartIndex: startIdx,
			EndIndex:   len(markdown),
		}
	}

	// Close last heading.
	if current != nil {
		current.Content = strings.TrimSpace(markdown[current.StartIndex:])
		current.EndIndex = len(markdown)
		headings = append(headings, *current)
	}

	return headings
}

// ChunkDocument splits a document into chunks by headings (port of TS chunkDocument).
// If tok is non-nil, it is used for accurate token counting; otherwise EstimateTokens is used.
func ChunkDocument(content string, path, title, description string, maxTokens int, tok Tokenizer) ChunkResult {
	if maxTokens <= 0 {
		maxTokens = 512
	}

	var chunks []Chunk
	totalTokens := 0

	// Chunk 0: metadata (title + description).
	metaContent := title
	if description != "" {
		metaContent = title + "\n\n" + description
	}
	metaTokens := countTokens(metaContent, tok)
	chunks = append(chunks, Chunk{
		ID:           fmt.Sprintf("doc:%s:chunk:0", path),
		Type:         ChunkTypeDoc,
		DocPath:      path,
		Section:      "# Metadata",
		Content:      metaContent,
		TokenCount:   metaTokens,
		HeadingLevel: 1,
		Position:     0,
	})
	totalTokens += metaTokens

	// Extract headings and chunk by section.
	headings := extractHeadings(content)
	position := 1

	// Capture H1 body content (between H1 and first H2) into metadata chunk.
	for _, h := range headings {
		if h.Level == 1 && strings.TrimSpace(h.Content) != "" {
			metaContent += "\n\n" + strings.TrimSpace(h.Content)
			break // only first H1
		}
	}
	// Update metadata chunk with any H1 body content.
	metaTokens = countTokens(metaContent, tok)
	chunks[0].Content = metaContent
	chunks[0].TokenCount = metaTokens

	for _, h := range headings {
		if h.Level == 1 {
			continue // skip h1 (title + body already in metadata chunk)
		}

		sectionTitle := fmt.Sprintf("%s %s", strings.Repeat("#", h.Level), h.Title)
		sectionContent := sectionTitle + "\n\n" + h.Content
		tokenCount := countTokens(sectionContent, tok)

		if tokenCount > maxTokens {
			// Split by paragraphs.
			paragraphs := splitParagraphs(h.Content)
			currentContent := sectionTitle
			currentTokens := countTokens(sectionTitle, tok)

			for _, para := range paragraphs {
				paraTokens := countTokens(para, tok)

				if currentTokens+paraTokens > maxTokens && currentContent != sectionTitle {
					// Save current chunk.
					chunks = append(chunks, Chunk{
						ID:           fmt.Sprintf("doc:%s:chunk:%d", path, position),
						Type:         ChunkTypeDoc,
						DocPath:      path,
						Section:      sectionTitle,
						Content:      currentContent,
						TokenCount:   currentTokens,
						HeadingLevel: h.Level,
						HeaderPath:   h.HeaderPath,
						Position:     position,
					})
					totalTokens += currentTokens
					position++

					// Start continuation chunk.
					currentContent = sectionTitle + " (continued)\n\n" + para
					currentTokens = countTokens(currentContent, tok)
				} else {
					currentContent += "\n\n" + para
					currentTokens += paraTokens
				}
			}

			// Remaining content.
			if currentContent != sectionTitle {
				chunks = append(chunks, Chunk{
					ID:           fmt.Sprintf("doc:%s:chunk:%d", path, position),
					Type:         ChunkTypeDoc,
					DocPath:      path,
					Section:      sectionTitle,
					Content:      currentContent,
					TokenCount:   currentTokens,
					HeadingLevel: h.Level,
					HeaderPath:   h.HeaderPath,
					Position:     position,
				})
				totalTokens += currentTokens
				position++
			}
		} else {
			// Fits in one chunk.
			chunks = append(chunks, Chunk{
				ID:           fmt.Sprintf("doc:%s:chunk:%d", path, position),
				Type:         ChunkTypeDoc,
				DocPath:      path,
				Section:      sectionTitle,
				Content:      sectionContent,
				TokenCount:   tokenCount,
				HeadingLevel: h.Level,
				HeaderPath:   h.HeaderPath,
				Position:     position,
			})
			totalTokens += tokenCount
			position++
		}
	}

	// If no headings, treat entire content as one chunk.
	if len(headings) == 0 && strings.TrimSpace(content) != "" {
		ct := countTokens(content, tok)
		chunks = append(chunks, Chunk{
			ID:           fmt.Sprintf("doc:%s:chunk:1", path),
			Type:         ChunkTypeDoc,
			DocPath:      path,
			Section:      "# Content",
			Content:      content,
			TokenCount:   ct,
			HeadingLevel: 1,
			Position:     1,
		})
		totalTokens += ct
	}

	return ChunkResult{Chunks: chunks, TotalTokens: totalTokens}
}

// ChunkTask splits a task into chunks by field (port of TS chunkTask).
// If tok is non-nil, it is used for accurate token counting; otherwise EstimateTokens is used.
// Fields exceeding maxTokens are split by paragraph.
func ChunkTask(task *models.Task, maxTokens int, tok Tokenizer) ChunkResult {
	if maxTokens <= 0 {
		maxTokens = 512
	}

	var chunks []Chunk
	totalTokens := 0

	type fieldSpec struct {
		Name    string
		Content string
	}

	fields := []fieldSpec{
		{"description", taskFieldContent(task, "description")},
		{"ac", taskFieldContent(task, "ac")},
		{"plan", taskFieldContent(task, "plan")},
		{"notes", taskFieldContent(task, "notes")},
	}

	for _, f := range fields {
		if f.Content == "" {
			continue
		}
		tc := countTokens(f.Content, tok)

		if tc > maxTokens {
			// Split by paragraphs, matching ChunkDocument behavior.
			paragraphs := splitParagraphs(f.Content)
			partIdx := 0
			currentContent := ""
			currentTokens := 0

			for _, para := range paragraphs {
				paraTokens := countTokens(para, tok)

				if currentTokens+paraTokens > maxTokens && currentContent != "" {
					// Save current chunk.
					suffix := ""
					if partIdx > 0 {
						suffix = fmt.Sprintf(":%d", partIdx)
					}
					chunks = append(chunks, Chunk{
						ID:         fmt.Sprintf("task:%s:chunk:%s%s", task.ID, f.Name, suffix),
						Type:       ChunkTypeTask,
						TaskID:     task.ID,
						Field:      f.Name,
						Content:    currentContent,
						TokenCount: currentTokens,
						Status:     task.Status,
						Priority:   task.Priority,
						Labels:     task.Labels,
					})
					totalTokens += currentTokens
					partIdx++

					currentContent = para
					currentTokens = paraTokens
				} else {
					if currentContent != "" {
						currentContent += "\n\n" + para
					} else {
						currentContent = para
					}
					currentTokens += paraTokens
				}
			}

			// Remaining content.
			if currentContent != "" {
				suffix := ""
				if partIdx > 0 {
					suffix = fmt.Sprintf(":%d", partIdx)
				}
				chunks = append(chunks, Chunk{
					ID:         fmt.Sprintf("task:%s:chunk:%s%s", task.ID, f.Name, suffix),
					Type:       ChunkTypeTask,
					TaskID:     task.ID,
					Field:      f.Name,
					Content:    currentContent,
					TokenCount: currentTokens,
					Status:     task.Status,
					Priority:   task.Priority,
					Labels:     task.Labels,
				})
				totalTokens += currentTokens
			}
		} else {
			chunks = append(chunks, Chunk{
				ID:         fmt.Sprintf("task:%s:chunk:%s", task.ID, f.Name),
				Type:       ChunkTypeTask,
				TaskID:     task.ID,
				Field:      f.Name,
				Content:    f.Content,
				TokenCount: tc,
				Status:     task.Status,
				Priority:   task.Priority,
				Labels:     task.Labels,
			})
			totalTokens += tc
		}
	}

	return ChunkResult{Chunks: chunks, TotalTokens: totalTokens}
}

// taskFieldContent extracts content for a specific field of a task.
func taskFieldContent(task *models.Task, field string) string {
	switch field {
	case "description":
		if task.Description != "" {
			return task.Title + "\n\n" + task.Description
		}
		return task.Title
	case "ac":
		if len(task.AcceptanceCriteria) == 0 {
			return ""
		}
		var parts []string
		for _, ac := range task.AcceptanceCriteria {
			check := "[ ]"
			if ac.Completed {
				check = "[x]"
			}
			parts = append(parts, fmt.Sprintf("- %s %s", check, ac.Text))
		}
		return strings.Join(parts, "\n")
	case "plan":
		return task.ImplementationPlan
	case "notes":
		return task.ImplementationNotes
	}
	return ""
}

var paragraphSplitRE = regexp.MustCompile(`\n\n+`)

func splitParagraphs(text string) []string {
	parts := paragraphSplitRE.Split(text, -1)
	var result []string
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			result = append(result, p)
		}
	}
	return result
}
